browser/brave_vpn/win/brave_vpn_wireguard_service/notifications/notification_utils.cc

```diff
@@ -0,0 +1,138 @@
+      RuntimeClass_Windows_Data_Xml_Dom_XmlDocument);
+  mswr::ComPtr<IInspectable> inspectable;
+  HRESULT hr =
+      base::win::RoActivateInstance(ref_class_name.get(), &inspectable);
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  mswr::ComPtr<winxml::Dom::IXmlDocumentIO> document_io;
+  hr = inspectable.As(&document_io);
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  base::win::ScopedHString ref_template =
+      base::win::ScopedHString::Create(xml_template);
+  hr = document_io->LoadXml(ref_template.get());
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  mswr::ComPtr<winxml::Dom::IXmlDocument> document;
+  hr = document_io.As<winxml::Dom::IXmlDocument>(&document);
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  mswr::ComPtr<winui::Notifications::IToastNotificationFactory>
+      toast_notification_factory;
+  hr = CreateActivationFactory(
+      RuntimeClass_Windows_UI_Notifications_ToastNotification,
+      IID_PPV_ARGS(&toast_notification_factory));
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+
+  mswr::ComPtr<winui::Notifications::IToastNotification> toast_notification;
+  hr = toast_notification_factory->CreateToastNotification(document.Get(),
+                                                           &toast_notification);
+  if (FAILED(hr)) {
+    return nullptr;
+  }
+  return toast_notification;
+}
+
+void ShowDesktopNotification(const std::wstring& text) {
+  base::win::AssertComInitialized();
+
+  mswr::ComPtr<winui::Notifications::IToastNotificationManagerStatics>
+      toast_manager;
+  HRESULT hr = CreateActivationFactory(
+      RuntimeClass_Windows_UI_Notifications_ToastNotificationManager,
+      IID_PPV_ARGS(&toast_manager));
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to create ToastNotificationManager, code: " << std::hex
+            << hr;
+    return;
+  }
+
+  mswr::ComPtr<winui::Notifications::IToastNotifier> notifier;
+  base::win::ScopedHString application_id =
+      base::win::ScopedHString::Create(install_static::GetBaseAppId());
+  hr =
+      toast_manager->CreateToastNotifierWithId(application_id.get(), &notifier);
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to create IToastNotifier, code: " << std::hex << hr;
+    return;
+  }
+
+  winui::Notifications::NotificationSetting setting;
+  hr = notifier->get_Setting(&setting);
+  if (setting != ABI::Windows::UI::Notifications::NotificationSetting_Enabled) {
+    VLOG(1) << "Notifications disabled for app";
+    return;
+  }
+
+  std::wstring content(kNotificationTemplate);
+  base::ReplaceSubstringsAfterOffset(&content, 0, L"{text}", text);
+  mswr::ComPtr<winui::Notifications::IToastNotification> toast =
+      GetToastNotification(content);
+  if (!toast) {
+    VLOG(1) << "Failed to create IToastNotification";
+    return;
+  }
+  hr = notifier->Show(toast.Get());
+  if (FAILED(hr)) {
+    VLOG(1) << "Failed to create IToastNotification, code: " << std::hex << hr;
+    return;
```

Comment:
This return is not needed. 

BTW, you can do the following pattern to avoid multiple VLOG:

```
base::expected<void, std::string> ShowDesktopNotificationImpl() {
    ...
  if (FAILED(hr)) {
     return base::unexpected("error message");
  }
  return base::ok();
}

void ShowDesktopNotification() {
 auto result = ShowDesktopNotificationImpl();
  if (!result.has_value()) {
    VLOG(1) << result.error();
  }
}
```

Reply:
done